versioned/storage/cache/src/test/java/org/projectnessie/versioned/storage/cache/TestCacheOvershoot.java

```diff
@@ -0,0 +1,131 @@
+  public void testCacheOvershootDelayedEviction() throws Exception {
+    // Production uses Runnable::run, but that lets this test sometimes run way too
+    // long, so we introduce some delay to simulate the case that eviction cannot keep up.
+    testCacheOvershoot(t -> delayedExecutor(2, TimeUnit.MILLISECONDS).execute(t));
+  }
+
+  private void testCacheOvershoot(Executor evictionExecutor) throws Exception {
+    var meterRegistry = new SimpleMeterRegistry();
+
+    var config =
+        CacheConfig.builder()
+            .capacityMb(4)
+            .cacheCapacityOvershoot(0.1d)
+            .executor(evictionExecutor)
+            .meterRegistry(meterRegistry)
+            .build();
+    var cache = new CaffeineCacheBackend(config);
+
+    var metersByName =
+        meterRegistry.getMeters().stream()
+            .collect(Collectors.toMap(m -> m.getId().getName(), Function.identity(), (a, b) -> a));
+    soft.assertThat(metersByName)
+        .containsKeys(METER_CACHE_WEIGHT, METER_CACHE_ADMIT_CAPACITY, METER_CACHE_REJECTED_WEIGHT);
+    var meterWeightReported = (Gauge) metersByName.get(METER_CACHE_WEIGHT);
+    var meterAdmittedCapacity = (Gauge) metersByName.get(METER_CACHE_ADMIT_CAPACITY);
+    var meterCapacity = (Gauge) metersByName.get(METER_CACHE_CAPACITY);
+    var meterRejectedWeight = (DistributionSummary) metersByName.get(METER_CACHE_REJECTED_WEIGHT);
+
+    var maxWeight = config.capacityMb() * 1024L * 1024L;
+    var admitWeight = cache.admitWeight();
+
+    var str = Strings.repeat("a", 4096);
+
+    var numThreads = 8;
+
+    for (int i = 0; i < maxWeight / 5000; i++) {
+      cache.put("repo", SimpleTestObj.builder().id(randomObjId()).text(str).build());
+    }
+
+    soft.assertThat(cache.currentWeightReported()).isLessThanOrEqualTo(admitWeight);
+    soft.assertThat(cache.rejections()).isEqualTo(0L);
+    soft.assertThat(meterWeightReported.value()).isGreaterThan(0d);
+    soft.assertThat(meterAdmittedCapacity.value()).isEqualTo((double) admitWeight);
+    soft.assertThat(meterCapacity.value()).isEqualTo((double) config.capacityMb() * ONE_MB);
+
+    var executor = Executors.newFixedThreadPool(numThreads);
+    var seenOvershoot = false;
+    var stop = new AtomicBoolean();
+    try {
+      for (int i = 0; i < numThreads; i++) {
+        executor.execute(
+            () -> {
+              while (!stop.get()) {
+                cache.put("repo", SimpleTestObj.builder().id(randomObjId()).text(str).build());
+                Thread.yield();
+              }
+            });
+      }
+
+      for (int i = 0; i < 50 && !seenOvershoot; i++) {
+        Thread.sleep(10);
+        var w = cache.currentWeightReported();
+        if (w > maxWeight) {
+          seenOvershoot = true;
+        }
+      }
+    } finally {
+      stop.set(true);
+
+      executor.shutdown();
+      executor.awaitTermination(10, TimeUnit.MINUTES);
+    }
+
+    soft.assertThat(cache.currentWeightReported()).isLessThanOrEqualTo(admitWeight);
```

Comment:
Minor: I believe after all insertions are done, current weight should not exceed the lower limit (no overshoot), because the executor performs eviction tasks immediately.